File: backend/models/hackathon.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Hackathon 活动表
type Hackathon struct {
	ID           uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	Name         string         `gorm:"type:varchar(100);not null" json:"name"`
	Description  string         `gorm:"type:text;not null" json:"description"`
	StartTime    time.Time      `gorm:"not null" json:"start_time"`
	EndTime      time.Time      `gorm:"not null" json:"end_time"`
	LocationType string         `gorm:"type:enum('online','offline','hybrid');not null" json:"location_type"`
	City         string         `gorm:"type:varchar(100)" json:"city"` // 城市
	LocationDetail string       `gorm:"type:varchar(500)" json:"location_detail"` // 具体地址
	Status       string         `gorm:"type:enum('preparation','published','registration','checkin','team_formation','submission','voting','results');default:'preparation'" json:"status"`
	OrganizerID  uint64         `gorm:"index;not null" json:"organizer_id"`
	MaxTeamSize  int            `gorm:"default:3" json:"max_team_size"`
	MaxParticipants int         `gorm:"default:0" json:"max_participants"` // 最大参与人数，0表示不限制
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联关系
	Organizer    User            `gorm:"foreignKey:OrganizerID" json:"organizer,omitempty"`
	Stages       []HackathonStage `gorm:"foreignKey:HackathonID" json:"stages,omitempty"`
	Awards       []HackathonAward `gorm:"foreignKey:HackathonID" json:"awards,omitempty"`
}

// TableName 指定表名
func (Hackathon) TableName() string {
	return "hackathons"
}

// HackathonStage 活动阶段时间表
type HackathonStage struct {
	ID          uint64    `gorm:"primaryKey;autoIncrement" json:"id"`
	HackathonID uint64    `gorm:"uniqueIndex:uk_hackathon_stage;not null" json:"hackathon_id"`
	Stage       string    `gorm:"uniqueIndex:uk_hackathon_stage;type:enum('registration','checkin','team_formation','submission','voting');not null" json:"stage"`
	StartTime   time.Time `gorm:"not null" json:"start_time"`
	EndTime     time.Time `gorm:"not null" json:"end_time"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName 指定表名
func (HackathonStage) TableName() string {
	return "hackathon_stages"
}

// HackathonAward 活动奖项表
type HackathonAward struct {
	ID          uint64    `gorm:"primaryKey;autoIncrement" json:"id"`
	HackathonID uint64    `gorm:"index;not null" json:"hackathon_id"`
	Name        string    `gorm:"type:varchar(100);not null" json:"name"`
	Prize       string    `gorm:"type:varchar(255);not null" json:"prize"`
	Quantity    int       `gorm:"default:1" json:"quantity"`
	Rank        int       `gorm:"not null" json:"rank"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName 指定表名
func (HackathonAward) TableName() string {
	return "hackathon_awards"
}

File: frontend/admin/src/api/hackathon.ts
import request from './request'

export interface Hackathon {
  id: number
  name: string
  description?: string
  status: 'preparation' | 'published' | 'registration' | 'checkin' | 'team_formation' | 'submission' | 'voting' | 'results' | 'cancelled'
  start_time: string
  end_time: string
  registration_start_time?: string
  registration_end_time?: string
  max_participants?: number
  current_participants?: number
  creator_id: number
  creator_name?: string
  organizer_id: number
  organizer_name?: string
  location_type: 'online' | 'offline' | 'hybrid'
  city?: string
  location_detail?: string
  max_team_size: number
  chain_event_id?: number
  created_at: string
  updated_at: string
}

export interface ChainData {
  event_id: number
  event_name: string
  description: string
  start_time: number
  end_time: number
  location: string
  organizer: string
  created_at: number
  updated_at: number
  is_deleted: boolean
}

export interface HackathonWithChainData {
  hackathon: Hackathon
  chain_data: ChainData | null
  chain_status: 'not_on_chain' | 'blockchain_error' | 'chain_data_error' | 'synced'
}

export interface HackathonListParams {
  page?: number
  page_size?: number
  status?: string
  keyword?: string
}

export interface HackathonListResponse {
  items: Hackathon[]
  total: number
  page: number
  page_size: number
}

export const getHackathonList = (params: HackathonListParams) => {
  return request.get<HackathonListResponse>('/hackathons', { params })
}

export const getHackathonDetail = (id: number) => {
  return request.get<Hackathon>(`/hackathons/${id}`)
}

export const getHackathonWithChainData = (id: number) => {
  return request.get<HackathonWithChainData>(`/hackathons/${id}/chain-data`)
}

export const verifyHackathonIntegrity = (id: number) => {
  return request.get(`/hackathons/${id}/verify`)
}

export const createHackathon = (data: Partial<Hackathon>) => {
  return request.post<Hackathon>('/hackathons', data)
}

export const updateHackathon = (id: number, data: Partial<Hackathon>) => {
  return request.put<Hackathon>(`/hackathons/${id}`, data)
}

export const deleteHackathon = (id: number) => {
  return request.delete(`/hackathons/${id}`)
}